File: leetcode/range-sum-query-immutable.cpp
#include <bits/stdc++.h>

using namespace std;

class NumArray {
private:
    vector<int> data;
public:
    NumArray(vector<int>& nums) {
        auto data_size = nums.size();
        data = nums;
        
        for(int i = 1; i < data_size; ++i) {
            data[i] += data[i-1];
        }
    }
    
    int sumRange(int left, int right) {
        int sub = left == 0 ? 0 : data[left-1];
        return data[right] - sub;
    }
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * int param_1 = obj->sumRange(left,right);
 */

File: leetcode/range-sum-query-immutable.py
from typing import List


class NumArray:
    def __init__(self, nums: List[int]):
        length = len(nums)
        self.presum = [nums[0]] * (length + 1)
        self.presum[0] = 0

        for i in range(1, length):
            self.presum[i + 1] = self.presum[i] + nums[i]

    def sumRange(self, left: int, right: int) -> int:
        return self.presum[right + 1] - self.presum[left]


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(left,right)
